src/torch_models.py

```diff
@@ -1,153 +1,323 @@
-        self._input_shift = torch.zeros(1)
-        self._input_scale = torch.zeros(1)
-        self._output_shift = torch.zeros(1)
-        self._output_scale = torch.zeros(1)
-        self._linears = nn.ModuleList()
-        self._loss_fn = nn.MSELoss()
 
-    def fit(self, X: Array, Y: Array) -> None:
-        """Train regressor on the given data.
+class PyTorchRegressor(Regressor, nn.Module):
+    """ABC for PyTorch regression models."""
+
+    def __init__(self, seed: int, max_train_iters: int, clip_gradients: bool,
+                 clip_value: float, learning_rate: float) -> None:
+        torch.manual_seed(seed)
+        Regressor.__init__(self, seed)
+        nn.Module.__init__(self)  # type: ignore
+        self._max_train_iters = max_train_iters
+        self._clip_gradients = clip_gradients
+        self._clip_value = clip_value
+        self._learning_rate = learning_rate
+
+    @abc.abstractmethod
+    def forward(self, tensor_X: Tensor) -> Tensor:
+        """Pytorch forward method."""
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def _initialize_net(self) -> None:
+        """Initialize the network once the data dimensions are known."""
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def _create_loss_fn(self) -> Callable[[Tensor, Tensor], Tensor]:
+        """Create the loss function used for optimization."""
+        raise NotImplementedError("Override me!")
+
+    def _create_optimizer(self) -> optim.Optimizer:
+        """Create an optimizer after the model is initialized."""
+        return optim.Adam(self.parameters(), lr=self._learning_rate)
+
+    def _fit(self, X: Array, Y: Array) -> None:
+        # Initialize the network.
+        self._initialize_net()
+        # Create the loss function.
+        loss_fn = self._create_loss_fn()
+        # Create the optimizer.
+        optimizer = self._create_optimizer()
+        # Convert data to tensors.
+        tensor_X = torch.from_numpy(np.array(X, dtype=np.float32))
+        tensor_Y = torch.from_numpy(np.array(Y, dtype=np.float32))
+        # Run training.
+        _train_predictive_pytorch_model(self,
+                                        loss_fn,
+                                        optimizer,
+                                        tensor_X,
+                                        tensor_Y,
+                                        max_iters=self._max_train_iters,
+                                        clip_gradients=self._clip_gradients,
+                                        clip_value=self._clip_value)
+
+    def _predict(self, x: Array) -> Array:
+        tensor_x = torch.from_numpy(np.array(x, dtype=np.float32))
+        tensor_X = tensor_x.unsqueeze(dim=0)
+        tensor_Y = self(tensor_X)
+        tensor_y = tensor_Y.squeeze(dim=0)
+        y = tensor_y.detach().numpy()  # type: ignore
+        return y
+
+
+class BinaryClassifier(abc.ABC):
```

Comment:
since there's non-torch stuff now, do we need to rename this file again? x.x (and the corresponding test file)

Reply:
i never liked the name models.py though, too generic